### 에러의 종류
1. 컴파일 오류
- 코드가 잘못되어서 문법적인 오류, 에러가 발생하는 것
- IDE 내에서 확인 가능
- 에러 핸들링에서는 다루지 않을 예정

2. 런타임 오류 
- 여러 다양한 이유로 앱이 실행되는 동안 발생하면 앱이 꺼지곤 한다.
- 미리 발생가능한 에러를 파악하고 올바르게 처리하면 앱이 꺼지는 걸 방지(크래쉬 방지)할 수 있다.
- 개발자가 처리해야 하는 에러가 이 종류다.

정보를 서버에 요청하고 데이터를 받아오는 통신을 하는데,
만약 데이터를 못 받아오면 어떻게 할래?
데이터 요청 함수는 실행이 제대로 되지 않아서 에러가 발생할 거야.

그동안 정의했던 함수를 '에러가 발생 가능한' 함수로 정의해보자.
비정상적인 경우 함수에서 에러를 '던질 수 있게' `throws` 키워드를 쓴다.

### 에러의 정의

에러는 **열거형**이다. 그래서 에러를 미리 정의해 두어야 한다.
애플이 미리 만들어둔 `Error Protocol`을 채택해야 한다.
프로토콜 채택을 통해 스위프트의 정식 에러 처리 문법을 따라 핸들링 가능하다.

```swift
enum Checking: Error {
    case aError
    case bError
    case cError
}

func checkError(a: Int) throws -> Bool {
    // 에러가 발생될 때 처리해야 한다면 throw 사용
    if a < 0 {
        throw Checking.aError
    }

    if a >= 7 {
        return true    
    } else {
        return false
    }
}

checkError(a: 3)

```
---
### 에러 처리의 3단계
1. 에러를 정의한다.
2. 에러가 발생할 수 있는 함수를 정의한다.
- 보통은 에러 코드 구문을 먼저 써주는 편인 것으로 보인다.
3. 에러 함수를 처리한다.
- `do`, `try`, `catch`
- 에러 핸들링은 아래처럼 한다.
    + do 블럭에서 try 함수 호출 -> 정상적인 경우에는 do 에서만 작동한다.
    + 그러니까 에러가 발생하지 않을 때는 do 블럭에서 처리된다.
    + 만약에 에러가 발생하면 catch블록으로 옮겨간다.
- 에러 타입 함수는 당연히 다른 타입과 구별할 수 있다.
- do-catch 블럭의 간소화된 형태도 있긴 하다.

```swift
do {
   var result = try checkError(a: 3) // do, try, catch     
   print(result) // false
} catch {
       print("에러 처리 구역")
}
```
---